docs/design/2023-06-30-configurable-kv-timeout.md

```diff
@@ -0,0 +1,238 @@
+                            false,
+                        );
+                        // Add timeout check.
+                        snap_store
+                        .get(&key, &mut statistics)
+                    });
+                 ...   
+            ...        
+        );
+}
+```
+These changes need to be done in the `tikv` repository.
+
+## Support Timeout Configuration For Coprocessor Tasks
+
+- Add related field in the `copTask` struct like
+```go
+type copTask struct {
+	...
+	KVReadTimeout: Duration
+}
+```
+- Pass the timeout value into the `Context.max_execution_duration_ms` field like
+```protobuf
+message Context {
+  ...
+  uint64 max_execution_duration_ms = 14;
+  ...
+}
+```
+
+- Use the `kv_read_timeout` value passed in to calculate the `deadline` result in `parse_and_handle_unary_request`,
+```rust
+fn parse_request_and_check_memory_locks(
+    &self,
+    mut req: coppb::Request,
+    peer: Option<String>,
+    is_streaming: bool,
+) -> Result<(RequestHandlerBuilder<E::Snap>, ReqContext)> {
+    ...
+     req_ctx = ReqContext::new(
+        tag,
+        context,
+        ranges,
+        self.max_handle_duration, // Here use the specified timeout value.
+        peer,
+        Some(is_desc_scan),
+        start_ts.into(),
+        cache_match_version,
+        self.perf_level,
+    );
+}
+```
+This change needs to be done in the `tikv` repository.
+
+## Timeout Retry
+
+- When timeout happens, retry the next available peer for the read requests like stale read and `non-leader-only`
+snapshot read requests. The strategy is:
+  1. Try the next peer if the first request times out immediately.
+  2. If all the responses from all the peers are `timeout` errors, return `timeout` error to the upper layer,
+  3. Backoff with the `timeout` error and retry the requests with **the orginal default** `ReadShortTimeout`
+  and `ReadMediumTimeout` values.
+  4. Continue to process like before.
+- The timeout retry details should be recorded and handled properly in the request tracker, so are the metrics.
+
+## Timeout Task Canceling
+
+- If the timeout value is configured to a small value like hundreds of milliseconds, there could be more
+timeout tasks. When task is spawned to the read pool in TiKV, it could not be canceled immediately for 
+serveral reasons:
+  - When the read task is polled and executed, there's no timeout checking mechanism in the task scheduler
+  by now.
+  - The read operations are synchronous, so if the read task is blocked by slow io the task could not be
+  scheduled or canceled.
+- By now some simple timeout checks could be added in certain places like "after getting snapshot". A complete
+  timeout check or scheduling design is needed in the future.
+
+
+## Compatibility
+
+The requests with configurable timeout values would take effect on newer versions. These fields are expected not to take effect when down-grading the
+cluster theoretically.
+
+## More comprehensive solution
```

Comment:
When we support cancelling a request from outside commands, we can implement a real hedge policies. For example
- For requests taking more than a threshold of time, send a hedge request to another replica. Either response can be taken as the final result.
- When either response is returned, cancel the other request.
- TiKV prioritizes non-hedge requests so the hedge policy won't increase latencies.

Reply:
This could be realized if the real `in-time canceling` is supported as described above.